### ISSUE TITLE
Add a Dependency tag system between mods.

### ISSUE THREAD

**Issue by Al12rs** [ModOrganizer2/modorganizer#838]
**The Problem**
There are many instances where some mods depend on others to work correctly, the easiest example is with patches of mods.
There is currently no way for the user to express these relations and no way for Mo2 to inform the user of incorrect configuration of the modlist.
In particular the problems we are looking to avoid are:
* Having dependency mods disabled while the dependent mod is active.
* Having mods in the incorrect order of dependencies, like a mod update should be located after the main mod to overwrite it's assets.
* Uninstalling a mod that is a dependency of another Mod.

**The solution**
Allow users to define dependencies between mods, so that MO2 may use that information to *inform* the user of an incorrect handling of the mods. Mo2 should only inform the user, not enforce automatically the application of the rule.

There would be different kinds of dependencies:

* Requires: this is the list of mods that the current mod needs to be enabled to function correctly.
* Is Required by: this is the list of mods that require the current mod to function correctly.

The creation of a Require dependency will automatically set the IsRequiredBy of the other mod and the other way around.

* Load after: this is the list of mods over which the current mod needs to have a higher priority. Load after does not imply Require, so if a mod in this list is missing then no error should be notified.
* Load Before: the list of mods of which current mod needs to have a lower priority. As before this does not imply Require relationship.

The setting of a Load After generates a Load Before in the other mod.
*Attention!* These types of rules can cause circular dependencies that would make the system always incorrect. So this brings a layer of complexity that is not present for simple Require relationships. A possible approach is to check the loadorder for circular dependencies when trying to add a new one and deny the operation in case it's destructive. As another option the Diagnose plugin mentioned later could also detect circular dependencies and inform the user of them.

A Diagnose plugin would then use the dependencies information of the mods to determine whether the current loadorder is correct and optionally propose an automatic fix for the issue.

An additional check would be useful when unistalling a mod, to see if it's required by some other mod.

**Implementing the feature:**
A system to define and remove dependencies would have to be designed, a possibility could be the use of a dedicated view accessible through right click to define dependencies, an alternative would be to add a dependencies tab to the ModInfoDialog.

Another matter is how to intuitively show dependencies between mods, which could be a second stage for the feature.
Things like flags for the various types of dependencies and coloring related mods when selecting one.

This feature could be implemented in stages.

**Additional Context:**
Vortex makes use of a dependency system, that was the inspiration of this feature suggestion, the main difference stays in the fact that Vortex tries to automatically enforce the rules set be the user, while Mo2 would instead leave the current ordering system intact and only inform the user of a possible discrepancy.

**Comment by mrudat** [comment]
Much of this is available on nexus mods already, presumably in a computer-readable format, in addition to the information stored in the .esps themselves. It is what is used to sort mods. Some of this information is returned if you attempt to sort mods where explicit dependencies are missing.

**Comment by Al12rs** [comment]
@mrudat I think you are misunderstanding, this would be for Mods priorities, not Plugins priorities. This would be for file overwriting conflicts between mods on the left pane, not plugins and record conflicts between those.

**Comment by selsie** [comment]
Loot does a pretty good job of that already, warning if a patch is missing.

**Comment by Kruziikrel13** [comment]
I think even a user managed form of this would have some incredible utility. Of course we've got LOOT for plugin management, but this kind of utility would be brilliant for management of mesh and texture mods for example. As these don't have any manner of detection for LOOT (naturally) We as user's would be able to set our own rules for our mods panel so that mod ordering can be handled in a more effective manner.

**Comment by twizz0r** [comment]
I'm not sure I see the utility here. MO already provides an elegant, intuitive way to see asset conflicts and an equally easy method to resolve them without having to reorder mods. 

LOOT is a great tool for plugins, but relies on human-generated information to do a lot of its work (and that information isn't always correct). 

Are you proposing a similar human-based information to manage left pane mod ordering?

I'm all for making modding easier for users, but the more rules you put in place, the more you obfuscate utility and also impose fences around what users can do.

**Comment by Kruziikrel13** [comment]
As the complexity of modlists grows it can be difficult for users to keep track of what why mods in their list overwrites others. Of course, we can make notes but that can also become somewhat verbose.

I think similar to loot, where when reading a mod's description if an overwrite order between two mods is explicitly described by a mod's description. It would be far easier for users to then carry this across as an "overwrite" / dependency rule that this feature addition would allow. 

Some file overwrites in MO2 can be meaningless as some mod authors incorporate the same files / meshes / scripts etc. in multiple of their mods which would of course then get flagged by MO2 as an overwrite.

Currently mod overwrite orders are implicitly correct... ie. there is room for ambiguity in regards to whether an overwrite matters or not. The dependency system would allow us to explicitly specify our overwrites.

Additionally, the dependency system would allow users to better keep track of what mods depend on what... I.e. mods that require Open Animation Replacer for Skyrim often don't include plugins that LOOT can raise issues about and exist purely as loose files. So setting these mods as dependent on OAR would be useful as, for example if OAR is then disabled the mods requiring it would then be able to notify the user one of their dependencies is missing immediately from MO2 rather than the user discovering this later when the mods don't work In game.

**Comment by twizz0r** [comment]
I'm half-joking, but you should see if you could train an 'ai" to do this. If it could crawl the web pages associated with URLs in meta files, conglomerate any install instructions (load after, hide files, etc.), look at established mod packs, and then make suggestions, that would do the trick.

Fixing Skyrim 's meshes is a maze. Lots of mods overwriting the same files with some files needing to be hidden so that assets lower in priority can win. If there was a tool to help put them in order, tell me what I'm missing and what I need to hide, that would be aces.

**Comment by Kruziikrel13** [comment]
Not sure I subscribe to the whole AI trend haha, It's far better at confidently lying than it is at actually providing information... User mesh management even with this system may still remain hell... but I think it may make it burn less? There are also of course other uses... a case I can think of off the top of my head would be TK Dodge and it's many derivatives updating it to 1.6.640 or NG'ing it or adding onto it etc.

**Comment by simonLeary42** [comment]
doesn't nexus mods already have all the dependencies? it gives me a popup saying what other mods I need to download first. does nexus have an API? can it be gathered with a bot?

**Comment by komlevv** [comment]
Any news or PRs on this? Have a bunch of custom mods and patches with patches of patches set up and as the list grows it's really hard to remember which one does what and why (end up just naming all the reqs in mod name). As the modlist grows its ez to accidentally delete something and break the whole system.

Package managers have this solved, is there any blocker to adapt existing models from Vortex/Nuget/etc?

**Comment by segln** [comment]
Mod authors know their mod dependencies best, and services like Nexus Mods already list them.
Instead of relying on external contributors to uncover hidden dependencies, it would be enough to simply check the author provided dependencies and trigger notifications.
